alpine/BumponTailInstabilityManager.h

```diff
@@ -0,0 +1,613 @@
+    }
+
+    void advance() override {
+        if (this->stepMethod_m == "LeapFrog") {
+            LeapFrogStep();
+        }
+    }
+
+    void LeapFrogStep(){
+        // LeapFrog time stepping https://en.wikipedia.org/wiki/Leapfrog_integration
+        // Here, we assume a constant charge-to-mass ratio of -1 for
+        // all the particles hence eliminating the need to store mass as
+        // an attribute
+        static IpplTimings::TimerRef PTimer           = IpplTimings::getTimer("pushVelocity");
+        static IpplTimings::TimerRef RTimer           = IpplTimings::getTimer("pushPosition");
+        static IpplTimings::TimerRef updateTimer      = IpplTimings::getTimer("update");
+        static IpplTimings::TimerRef domainDecomposition = IpplTimings::getTimer("loadBalance");
+        static IpplTimings::TimerRef SolveTimer       = IpplTimings::getTimer("solve");
+
+        double dt                               = dt_m;
+        std::shared_ptr<ParticleContainer_t> pc = pcontainer_m;
+        std::shared_ptr<FieldContainer_t> fc    = fcontainer_m;
+
+        IpplTimings::startTimer(PTimer);
+        pc->P = pc->P - 0.5 * dt * pc->E;
+        IpplTimings::stopTimer(PTimer);
+
+        // drift
+        IpplTimings::startTimer(RTimer);
+        pc->R = pc->R + dt * pc->P;
+        IpplTimings::stopTimer(RTimer);
+
+        // Since the particles have moved spatially update them to correct processors
+        IpplTimings::startTimer(updateTimer);
+        pc->update();
+        IpplTimings::stopTimer(updateTimer);
+
+        size_type totalP          = totalP_m;
+        int it                    = it_m;
+        bool isFirstRepartition = false;
+        if (loadbalancer_m->balance(totalP, it + 1)) {
+                IpplTimings::startTimer(domainDecomposition);
+                auto* mesh = &fc->getRho().get_mesh();
+                auto* FL = &fc->getFL();
+                loadbalancer_m->repartition(FL, mesh, isFirstRepartition);
+                IpplTimings::stopTimer(domainDecomposition);
+        }
+
+        // scatter the charge onto the underlying grid
+        this->par2grid();
+
+        // Field solve
+        IpplTimings::startTimer(SolveTimer);
+        this->fsolver_m->runSolver();
+        IpplTimings::stopTimer(SolveTimer);
+
+        // gather E field
+        this->grid2par();
+
+        // kick
+        IpplTimings::startTimer(PTimer);
+        pc->P = pc->P - 0.5 * dt * pc->E;
+        IpplTimings::stopTimer(PTimer);
+    }
+
+    void par2grid() override { scatterCIC(); }
+
+    void grid2par() override { gatherCIC(); }
+
+    void gatherCIC() {
+        using Base                       = ippl::ParticleBase<ippl::ParticleSpatialLayout<T, Dim>>;
+        Base::particle_position_type *Ep = &pcontainer_m->E;
+        Base::particle_position_type *R  = &pcontainer_m->R;
+        VField_t<T, Dim> *Ef             = &fcontainer_m->getE();
+        gather(*Ep, *Ef, *R);
+    }
```

Comment:
You mean an ALPINE-level class which is derived from`PicManager` and the individual managers like `LandauDampingManager` are then derived from it?

Reply:
Yes I agree we should avoid code repetition as much as possible..

Reply:
Yes, exactly. `ChargedParticles` contained all the common elements (and a few things it arguably shouldn't have contained) but otherwise we should employ the same strategy; a common manager class for all the ALPINE apps can hold the common elements.

Reply:
Something like this?
b0c0a86

Reply:
May be `AlpineManager` is a better name than `PicManagerCommon`.

Reply:
I agree regarding the naming, since this is specific to ALPINE. But yes, this is what I meant.

Reply:
done in 39d6f6d